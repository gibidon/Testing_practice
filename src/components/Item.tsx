import { validateHeaderMax, validateHeaderMin,HEADER_LENGTH_ERROR_MESSAGE } from "src/utils/helpers";
import { DeleteButton } from "./DeleteButton";

type Props = Task & {
  onDelete: (id: Task["id"]) => void;
  onToggle: (id: Task["id"]) => void;
};

export const Item = (props: Props) => {
  // const disabled = validateHeaderMax(props.header) || validateHeaderMin(props.header)
  const disabled = validateHeaderMax(props.header) && validateHeaderMin(props.header)

  return (
    <li className="item-wrapper">
      <input
        type="checkbox"
        id={props.id}
        defaultChecked={props.done}
        onChange={() => props.onToggle(props.id)}
      />
      <label htmlFor={props.id} onClick={() => props.onToggle(props.id)}>
        {/* {props.done ? <s>{props.header}</s> : props.header} */}
        {disabled ? props.header :HEADER_LENGTH_ERROR_MESSAGE}
      </label>
      <DeleteButton
        disabled={!props.done}
        onClick={() => props.onDelete(props.id)}
      />
    </li>
  );
};
