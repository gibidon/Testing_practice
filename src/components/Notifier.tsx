import { useEffect } from "react";

type Props = {
    task: string;
    open: boolean;
    onClose: () => void;
}

export const Notifier = ({ task, open, onClose }: Props) => {
    useEffect(() => {
        let timerId: NodeJS.Timeout;
        console.log('applying setTimeout in useEffect')
        if (open) {
            timerId = setTimeout(() => {
                onClose();
            }, 2000);
        }

        return () => {
            console.log('applying setTimeout in useEffect')

            clearTimeout(timerId)};
    }, [open]);

    if (!open) {
        return null;
    }

    return(
        <div className="blackout">
            <div className="notifier-wrapper" data-testid='notifier'>
                {task}
            </div>
        </div>
    );
}