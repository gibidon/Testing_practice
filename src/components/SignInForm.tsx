import { ChangeEvent, useState,useRef, FormEvent, FormEventHandler } from "react"

const initialInputState = {
    name:'',
    surname:'',
    age:0
}

export const SignInForm = () => {
    const [inputs,setInputs] = useState(initialInputState)
    const formRef = useRef<HTMLFormElement>(null)
    const inputNameRef = useRef<HTMLInputElement>(null)
    const handleChange:FormEventHandler = (e:ChangeEvent<HTMLInputElement>) => {
        setInputs((prevState) => ({...prevState,[e.target.name]:e.target.value
        }))
    }

    const handleSubmit = (e:FormEvent<HTMLFormElement>) => {
        e.preventDefault()
        console.log('inputs submitted: ',inputs )
        
        formRef.current?.reset()
    }

    const handleReset = () => {
        setInputs(initialInputState)
    }

    const handleBegin = () => {
        inputNameRef.current?.focus()
    }
    return (
        <>
        <p><button onClick={handleBegin}>Fill out the form</button></p>
         <form 
            ref={formRef}
            onChange={handleChange}
            onSubmit={handleSubmit}
            onReset={handleReset}
        >
            <input  type='text' name='name' ref={inputNameRef}/>
            <input  type='text' name='surname'/>
            <input  type='number' name='age'/>
            <p>
                <button type='submit'>Submit</button>
                <button type='reset'>Reset</button>
            </p>
        </form>
        </>
       
    )
}