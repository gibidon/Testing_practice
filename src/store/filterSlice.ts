import { createSlice,current } from "@reduxjs/toolkit";
import { PayloadAction } from "@reduxjs/toolkit";
import { RootState } from "./configureStore";

interface IFilters {
  // filters:{
    onlyActive:boolean
  }  
// }


const initialFilterParams:IFilters = {
  // filters:{
    onlyActive:false,
}
// }
export const filterSlice = createSlice({
    name:'filters',
    initialState:initialFilterParams,
    reducers:{
      toggleOnlyActive(state){
        state.onlyActive = !state.onlyActive
      },
      updateFilterParams(state,action:PayloadAction<{key:keyof IFilters,value:string | boolean}>){
      // updateFilterParams(state,action:PayloadAction<{key:string,value:string | boolean}>){
        console.log('act',action.payload)
        console.log('prevState',state)
        console.log('current:',current(state))
        const currentState = current(state)
        // state = {...state,[action.payload.key]:action.payload.value}
        // state[action.payload.key] = action.payload.value

        console.log('new state:',state)
      }
    }
})

export default filterSlice.reducer

export const {toggleOnlyActive,updateFilterParams} = filterSlice.actions 

export const isOnlyActive = (state:RootState) => state.filters.onlyActive