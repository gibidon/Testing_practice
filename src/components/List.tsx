import { Item } from "./Item";

type Props = {
  items: Task[];
  onDelete: (id: Task["id"]) => void;
  onToggle: (id: Task["id"]) => void;
};

export const List = ({ items, onDelete, onToggle }: Props) => {
  const itemsToRender = items.length <= 10 ? items : items.slice(0,9) 

  return (
  <ul className="task-list tasks">
    {/* {items.map((item) => ( */}
    {itemsToRender.map((item) => (
      <Item
        {...item}
        key={item.id}
        onDelete={onDelete}
        onToggle={onToggle} />
    ))}
  </ul>
)
}