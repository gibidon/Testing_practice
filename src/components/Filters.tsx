import { useState } from "react"
import { useDispatch, useSelector } from "react-redux"
import { RootState } from "src/store/configureStore"
import { toggleOnlyActive, updateFilterParams } from "src/store/filterSlice"

interface Filters {
    onlyFiltered:boolean
    
}

const initialFilters:Filters = {
    onlyFiltered:false
}

export const Filters = () => {
    

    return (<div>
        <Switch/>
    </div>)
    
}

const Switch = () => {
    const dispatch = useDispatch()
    const onlyActive = useSelector((state:RootState) =>state.filters.onlyActive)

    return <div>
        <button data-testid='onlyActiveTasksBtn' className="onlyActiveTasksBtn" onClick={() => dispatch(toggleOnlyActive())}>{onlyActive ? 'Show all' : 'Show only active'}</button>
        {/* <button onClick={() => dispatch(updateFilterParams({key:"onlyActive",value:!onlyActive}))}>{onlyActive ? 'Show all' : 'Show only active'}</button> */}
        {/* <button onClick={() => dispatch(updateFilterParams({key:"genre",value:genre ==='male' ? 'female' : 'male'}))}>{genre === 'male' ? 'female' : 'male'}</button> */}
    </div>
}