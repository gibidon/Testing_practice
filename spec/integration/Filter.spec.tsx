import { render,screen,prettyDOM } from "@testing-library/react";
import { App } from "src/App";
import { List } from "src/components/List";
import { JestStoreProvider } from "../utils/JestStoreProvider";
import ue from '@testing-library/user-event'
import * as taskSliceModule from 'src/store/taskSlice'

const tasksMock:Task[] = [
    {id:'0',header:'mock0',done:true},
    {id:'1',header:'mock1',done:false},
    {id:'2',header:'mock2',done:true},
    {id:'3',header:'mock3',done:true},
    {id:'4',header:'mock4',done:false},
]


jest.useRealTimers()

describe('Список задач', () => {
    const userEvent = ue.setup()   

    test.skip('с включенным фильтром не содержит выполненные задачи после нажатия на кнопку фильтрации',async() => {
        // const userEvent = ue.setup({advanceTimers:jest.advanceTimersByTime})    

        const spied = jest.spyOn(taskSliceModule,'tasksSelector')
        spied.mockReturnValue(tasksMock)

        const {rerender} = render(<App />,{wrapper:JestStoreProvider})

        const onlyActiveTasksBtn = screen.getByTestId('onlyActiveTasksBtn')
        await userEvent.click(onlyActiveTasksBtn)

        const taskList = screen.queryAllByRole('listitem')
        expect(taskList).toHaveLength(2)

        await userEvent.click(onlyActiveTasksBtn)
 
        rerender(<App/>)
        const newTaskList = screen.getAllByRole('listitem')
        expect(newTaskList).toHaveLength(5)        
    });
    test('с включенным фильтром не содержит выполненные задачи после нажатия на кнопку фильтрации-вариант через снапшот',async() => {

        const spied = jest.spyOn(taskSliceModule,'tasksSelector')
        spied.mockReturnValue(tasksMock)

        const onDelete = jest.fn()
        const onToggle = jest.fn()
        const {rerender,asFragment} = render(<List items={tasksMock} onDelete={onDelete} onToggle={onToggle}/>)

        const firstRender = asFragment()
        const filteredTasksMock = tasksMock.filter(task => task.done === false)

        // const onlyActiveTasksBtn = screen.getByTestId('onlyActiveTasksBtn')
        // await userEvent.click(onlyActiveTasksBtn)

        rerender(<List items={filteredTasksMock} onDelete={onDelete} onToggle={onToggle}/>)
        const secondRender = asFragment()
        
        
        // await userEvent.click(onlyActiveTasksBtn)
        

        expect(firstRender).toMatchDiffSnapshot(secondRender)
              
    });
    
    // it.todo('с выключенным фильтром');
});
