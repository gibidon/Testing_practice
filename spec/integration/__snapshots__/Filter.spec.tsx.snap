// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Список задач с включенным фильтром не содержит выполненные задачи после нажатия на кнопку фильтрации-вариант через снапшот 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -4,33 +4,10 @@
    >
      <li
        class="item-wrapper"
      >
        <input
-         checked=""
-         id="0"
-         type="checkbox"
-       />
-       <label
-         for="0"
-       >
-         mock0
-       </label>
-       <button
-         class="button button-with-icon"
-         data-alt="Удалить"
-       >
-         <img
-           alt="Удалить"
-           src="[object Object]"
-         />
-       </button>
-     </li>
-     <li
-       class="item-wrapper"
-     >
-       <input
          id="1"
          type="checkbox"
        />
        <label
          for="1"
@@ -39,56 +16,10 @@
        </label>
        <button
          class="button button-with-icon"
          data-alt="Удалить"
          disabled=""
-       >
-         <img
-           alt="Удалить"
-           src="[object Object]"
-         />
-       </button>
-     </li>
-     <li
-       class="item-wrapper"
-     >
-       <input
-         checked=""
-         id="2"
-         type="checkbox"
-       />
-       <label
-         for="2"
-       >
-         mock2
-       </label>
-       <button
-         class="button button-with-icon"
-         data-alt="Удалить"
-       >
-         <img
-           alt="Удалить"
-           src="[object Object]"
-         />
-       </button>
-     </li>
-     <li
-       class="item-wrapper"
-     >
-       <input
-         checked=""
-         id="3"
-         type="checkbox"
-       />
-       <label
-         for="3"
-       >
-         mock3
-       </label>
-       <button
-         class="button button-with-icon"
-         data-alt="Удалить"
        >
          <img
            alt="Удалить"
            src="[object Object]"
          />"
`;
